### ISSUE TITLE
TradingAccount Data Being Deleted After Partial Update

### ISSUE THREAD

**Issue by geoffo** [Sage/SData-2.0#14]
I am attempting to update trading accounts using a partial update payload, as stated on the following page: http://sage.github.io/SData-2.0/pages/core/0902/

However when I perform the update to change a single field, the rest of the fields are reset/deleted.

Here is the response payload from the initial POST (some nil fields removed for brevity):

```
<sdata:payload>
 <crm:tradingAccount sdata:url="http://localhost:5493/sdata/accounts50/GCRM/-/tradingAccountCustomer(b1c1e41d-c0bd-4461-8605-9a88b9609a15)?format=atomentry" sdata:uuid="b1c1e41d-c0bd-4461-8605-9a88b9609a15">
  <crm:active>true</crm:active>
  <crm:customerSupplierFlag>Customer</crm:customerSupplierFlag>
  <crm:companyPersonFlag>Company</crm:companyPersonFlag>
  <crm:invoiceTradingAccount xsi:nil="true" />
  <crm:openedDate>2015-05-10</crm:openedDate>
  <crm:reference>ZZTESTUP</crm:reference>
  <crm:reference2 xsi:nil="true" />
  <crm:status>Open</crm:status>
  <crm:name>ZZTestUpdateCust</crm:name>
  <crm:shortName xsi:nil="true" />
  <crm:type>Unknown</crm:type>
  <crm:website>www.test.com</crm:website>
  <crm:deliveryRule>true</crm:deliveryRule>
  <crm:currency>GBP</crm:currency>
  <crm:taxReference />
  <crm:taxationCountry>GB</crm:taxationCountry>
  <crm:financeBalance>100.00</crm:financeBalance>
  <crm:financeLimit>10.00</crm:financeLimit>
  <crm:financeStatusFlag>false</crm:financeStatusFlag>
  <crm:financeStatusText>Open</crm:financeStatusText>
  <crm:settlementDiscountType>Percent</crm:settlementDiscountType>
  <crm:settlementDiscountPercent>10.00</crm:settlementDiscountPercent>
  <crm:settlementDiscountTerms>45</crm:settlementDiscountTerms>
  <crm:settlementDiscountIncludedInTotal>true</crm:settlementDiscountIncludedInTotal>
  <crm:paymentTerms>11</crm:paymentTerms>
  <crm:orderLineDiscountPercent>0.00</crm:orderLineDiscountPercent>
  <crm:invoiceDiscountPercent>0.00</crm:invoiceDiscountPercent>
  <crm:primacyIndicator>true</crm:primacyIndicator>
  <crm:balance>100.00</crm:balance>
  <crm:creditLimit>10.00</crm:creditLimit>
  <crm:MTDTurnover>0.00</crm:MTDTurnover>
  <crm:YTDTurnover>0.00</crm:YTDTurnover>
  <crm:priorYTDTurnover>0.00</crm:priorYTDTurnover>
  <crm:accountOpenedDate>2015-05-10</crm:accountOpenedDate>
  <crm:tradingTerms>30 Days Strictly Net</crm:tradingTerms>
  <crm:standardDiscount>0.00</crm:standardDiscount>
  <crm:accountOnHold>false</crm:accountOnHold>
  <crm:analysis1>Private</crm:analysis1>
  <crm:analysis2>Jenkins</crm:analysis2>
  <crm:analysis3>Chapel</crm:analysis3>
  <crm:canChargeCredit>false</crm:canChargeCredit>
  <crm:additionalDiscount>0</crm:additionalDiscount>
 </crm:tradingAccount>
</sdata:payload>
```

Here is the PUT request to update the trading account:

```
PUT http://localhost:5493/sdata/accounts50/GCRM/-/tradingAccounts('b1c1e41d-c0bd-4461-8605-9a88b9609a15') HTTP/1.1
Host: localhost:5493
Authorization: Basic ********
Accept-Encoding: gzip, deflate
Content-Type: application/atom+xml; type=entry
Content-Length: 389

<entry xmlns:sdata="http://schemas.sage.com/sdata/2008/1" xmlns:crm="http://schemas.sage.com/crmErp/2008" xmlns:sc="http://schemas.sage.com/sc/2009" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.w3.org/2005/Atom">
  <sdata:payload>
    <crm:tradingAccount>
      <crm:analysis3>Analysis3</crm:analysis3>
    </crm:tradingAccount>
  </sdata:payload>
</entry>
```

Here is the response payload for the PUT request (some nil fields removed for brevity):

```
<sdata:payload>
 <crm:tradingAccount sdata:url="http://localhost:5493/sdata/accounts50/GCRM/-/TRADINGACCOUNTCUSTOMER(b1c1e41d-c0bd-4461-8605-9a88b9609a15)?format=atomentry" sdata:uuid="b1c1e41d-c0bd-4461-8605-9a88b9609a15">
  <crm:active>true</crm:active>
  <crm:customerSupplierFlag>Customer</crm:customerSupplierFlag>
  <crm:companyPersonFlag>Company</crm:companyPersonFlag>
  <crm:invoiceTradingAccount xsi:nil="true" />
  <crm:openedDate xsi:nil="true" />
  <crm:reference>ZZTESTUP</crm:reference>
  <crm:reference2 xsi:nil="true" />
  <crm:status>Open</crm:status>
  <crm:name />
  <crm:shortName xsi:nil="true" />
  <crm:type>Unknown</crm:type>
  <crm:website />
  <crm:deliveryRule>true</crm:deliveryRule>
  <crm:currency>GBP</crm:currency>
  <crm:taxCode xsi:nil="true" />
  <crm:taxReference />
  <crm:taxationCountry>GB</crm:taxationCountry>
  <crm:useTaxInclusivePrices xsi:nil="true" />
  <crm:financeBalance>0.00</crm:financeBalance>
  <crm:financeLimit>0.00</crm:financeLimit>
  <crm:financeBalanceDate xsi:nil="true" />
  <crm:financeStatusFlag>false</crm:financeStatusFlag>
  <crm:financeStatusText>Open</crm:financeStatusText>
  <crm:settlementDiscountType>Percent</crm:settlementDiscountType>
  <crm:settlementDiscountAmount xsi:nil="true" />
  <crm:settlementDiscountPercent>0.00</crm:settlementDiscountPercent>
  <crm:settlementDiscountTerms>0</crm:settlementDiscountTerms>
  <crm:settlementDiscountTermsCommencement xsi:nil="true" />
  <crm:settlementDiscountIncludedInTotal>true</crm:settlementDiscountIncludedInTotal>
  <crm:paymentTerms>0</crm:paymentTerms>
  <crm:orderLineDiscountType xsi:nil="true" />
  <crm:orderLineDiscountAmount xsi:nil="true" />
  <crm:orderLineDiscountPercent>0.00</crm:orderLineDiscountPercent>
  <crm:invoiceDiscountType xsi:nil="true" />
  <crm:invoiceDiscountAmount xsi:nil="true" />
  <crm:invoiceDiscountPercent>0.00</crm:invoiceDiscountPercent>
  <crm:defaultFulfillmentLocation xsi:nil="true" />
  <crm:primacyIndicator>true</crm:primacyIndicator>
  <crm:balance>0.00</crm:balance>
  <crm:creditLimit>0.00</crm:creditLimit>
  <crm:MTDTurnover>0.00</crm:MTDTurnover>
  <crm:YTDTurnover>0.00</crm:YTDTurnover>
  <crm:priorYTDTurnover>0.00</crm:priorYTDTurnover>
  <crm:tradingTerms>30 Days Strictly Net</crm:tradingTerms>
  <crm:standardDiscount>0.00</crm:standardDiscount>
  <crm:accountOnHold>false</crm:accountOnHold>
  <crm:analysis1 />
  <crm:analysis2 />
  <crm:analysis3>Analysis3</crm:analysis3>
  <crm:canChargeCredit>false</crm:canChargeCredit>
  <crm:additionalDiscount>0</crm:additionalDiscount>
 </crm:tradingAccount>
</sdata:payload>
```

As you can see in the above payloads, after the update, the name, website, balances, etc. have all been reset.  In my testing here, this only appears to be happening on trading accounts (as some other tables I have been able to perform partial updates).

**Comment by darroncockram** [comment]
Not seeing that behaviour here. If I do a partial update and only change the analysis3 field all other fields correctly retain their current values. What version of Accounts are you testing against?

**Comment by geoffo** [comment]
Hmmm... I am using Sage 50 Accounts Professional 2015.  Here are the product details from the about page:

Version Number: 21.3.6.108
SDB Desktop Version: 12.1.369.0
Report Designer Version: 1.4.2013.27
Updates Installed: SDO Update 1, Service Pack 4

I was actually thinking that something may be out of date since the issues I am seeing are working for you.  Can you confirm if something on my end needs to be updated and if so how I should go about updating?

**Comment by darroncockram** [comment]
You are up to date and I'm using the same version as you although I'd also tested it with Accounts 2014 and 2013 with the same results.
I have spotted where our two approaches differ though. When forming my PUT request I use the sdata:url attribute of the entry returned from my initial GET request as the URL for the PUT, whereas you are building it directly. There is a key difference here as your URL is:

http://localhost:5493/sdata/Accounts50/GCRM/-/__tradingAccounts__('b1c1e41d-c0bd-4461-8605-9a88b9609a15')

but mine is:

http://localhost:5493/sdata/Accounts50/GCRM/-/__tradingAccountCustomer__(b1c1e41d-c0bd-4461-8605-9a88b9609a15)

Technically both are correct but there is a difference in how they are processed behind the scenes and using the former does result in the behaviour you describe. Using the latter does not and is why I could not initially replicate your results.
I've raised this with the Accounts team, although I suspect this is a problem with the SIF rather than Accounts. For now the workaround would be to use the latter approach (if you are using the SIF it's the Uri property of the feedEntry that you want to use. If not it's simply the sdata:url attribute of the crm:tradingAccount element from the GET request that you want.

Hope that helps.
